### PR TITLE
Fix EZP-24753: non-admin load user on login to PlatformUI

### PR DESCRIPTION
**Pull request by joaoinacio**
JIRA: https://jira.ez.no/browse/EZP-24753

### CODE REVIEW

File: eZ/Publish/Core/REST/Server/Controller/User.php [hunk 1]
```diff
@@ -179,16 +179,31 @@ public function loadUser($userId)
         $user = $this->userService->loadUser($userId);
 
         $userContentInfo = $user->getVersionInfo()->getContentInfo();
-        $userMainLocation = $this->locationService->loadLocation($userContentInfo->mainLocationId);
         $contentType = $this->contentTypeService->loadContentType($userContentInfo->contentTypeId);
 
+        $mainLocationId = $userContentInfo->mainLocationId;
+
+        // special case for loading current user, which does not have permissions to load its location.
+        if ($user == $this->repository->getCurrentUser()) {
```
Comment (andrerom) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: this condition will always be true, there is always a user.

Comment (joaoinacio) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: @andrerom : I'm not sure I understand.

$user will be the user which is requested for the REST request;
currentUser should be the current logged in user, or anonymous.

With the above, if user with ID 50 tries to load f.e. user with ID 10, the condition will be false?

Comment (andrerom) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: Ah, I didn't see the `==`, thought you where doing a `=`, please make the sentance more clear by doing:
`if ($user->id === $this->repository->getCurrentUser()->id) {`

File: eZ/Publish/Core/REST/Server/Controller/User.php [hunk 2]
```diff
@@ -179,16 +179,31 @@ public function loadUser($userId)
...
+            $locationService = $this->locationService;
+            $userMainLocation = $this->repository->sudo(
+                function () use ($locationService, $mainLocationId) {
+                    return $locationService->loadLocation($mainLocationId);
+                }
+            );
+            $relations = array();
```
Comment (lolautruche) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: Why do you leave `$relations` empty here?

File: eZ/Publish/Core/REST/Server/Controller/User.php (hunk 1 above)
Comment (lolautruche) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: +1 with @andrerom

File: eZ/Publish/Core/REST/Server/Controller/User.php (hunk 2 above)
Comment (joaoinacio) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: Not really certain how relations should be handled.

If loading the own user's location shouldn't pose any problem, using sudo to load relations could expose restricted content, hence a potential security issue?

Comment (lolautruche) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: Right, +1 for leaving it empty.

File: eZ/Publish/Core/REST/Server/Controller/User.php [hunk 3]
```diff
@@ -179,16 +179,31 @@ public function loadUser($userId)
...
 
         $userContentInfo = $user->getVersionInfo()->getContentInfo();
-        $userMainLocation = $this->locationService->loadLocation($userContentInfo->mainLocationId);
         $contentType = $this->contentTypeService->loadContentType($userContentInfo->contentTypeId);
 
+        $mainLocationId = $userContentInfo->mainLocationId;
+
+        // special case for loading current user, which does not have permissions to load its location.
+        if ($user->id == $this->repository->getCurrentUser()->id) {
+            $locationService = $this->locationService;
```
Comment (lolautruche) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: This is not needed any more as we require PHP 5.4 as of eZ 5.4, unless we want to backport this to 5.3.

**Comment by lolautruche** [comment]
+1 Besides my last nitpick.
Travis failure doesn't seem to be related.

Any idea @andrerom @vidarl ?

Ping @bdunogier @dpobel @yannickroger @glye

**Comment by lolautruche** [comment]
And of course ping @pspanja

File: eZ/Publish/Core/REST/Server/Controller/User.php (hunk 2 above)
Comment (andrerom) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: please add inline comment on this

File: eZ/Publish/Core/REST/Server/Controller/User.php (hunk 3 above)
Comment (andrerom) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: Are you implying binding $this to the closure?

**Comment by pspanja** [comment]
+1 with last remark on closure resolved.

File: eZ/Publish/Core/REST/Server/Controller/User.php (hunk 3 above)
Comment (lolautruche) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: Yes, `$this` is correctly dereferenced in closures as of PHP 5.4

File: eZ/Publish/Core/REST/Server/Controller/User.php (hunk 2 above)
Comment (andrerom) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: /cc @bdunogier on this ;)

**Comment by joaoinacio** [comment]
Updated according to feedback, also modified the flow a bit (sorry!) to only try sudo if the first load fails.
This allows f.e. relations to be loaded for admin user, etc - only if it fails is sudo attemped.
(AFAIK there should be no perf considerations on login).

If needed the change can be reverted, sorry again and thanks for the review

**Comment by andrerom** [comment]
@joaoinacio please don't change things like that after review, that makes the review worthless.
It looks slightly more of a hack now, but as it is hack anyway, so +1 BUT:

please change your comment like this, and no other line:
```diff
- // special case: allow current logged in user to load location for its own User content
+ // TODO: Hack for special case to allow current logged in user to load him/here self (but not relations)
```

**Comment by lolautruche** [comment]
+1

**Comment by joaoinacio** [comment]
Done.

**Comment by bdunogier** [comment]
Hack indeed, but let's say +1.

It would be nice to cover in BDD that non-admin editors should be able to login to PlatformUI, even if it is as a follow-up.

**Comment by andrerom** [comment]
> It would be nice to cover in BDD that non-admin editors should be able to login to PlatformUI, even if it is as a follow-up.

very good point, @miguelcleverti could that be the QA part on this story maybe? *(instead of manual tests)*

**Comment by andrerom** [comment]
Merging, thanks @joaoinacio !

**Pull request merged**